File: packages/ios/OpenCode/Views/ToolDisplay.swift
import Foundation

/// Compact, web-client-style presentation of a message's non-text parts: a verb
/// label plus a concise target (file / command / pattern) and an optional badge
/// (diff counts, match count). Pure value logic so it can be unit-tested against
/// real session data, mirroring `getToolInfo()` in the OpenCode web client
/// (packages/ui/src/components/message-part.tsx).
enum ToolDisplay {
    /// The verb shown for a tool, matching the web client's English labels.
    static func label(for tool: String) -> String {
        switch tool.lowercased() {
        case "read":               return "Read"
        case "list":               return "List"
        case "glob":               return "Glob"
        case "grep":               return "Grep"
        case "bash", "shell":      return "Shell"
        case "edit":               return "Edit"
        case "write":              return "Write"
        case "patch", "apply_patch": return "Patch"
        case "webfetch":           return "Webfetch"
        case "websearch":          return "Web Search"
        case "task":               return "Agent"
        case "todowrite", "todo":  return "To-dos"
        case "question":           return "Questions"
        default:                   return tool
        }
    }

    /// A tool's verb plus a one-line target + badge, e.g.
    /// `("Edit", "ast_query.h  +5 −4")`, `("Grep", "Foo|Bar  no matches")`.
    static func describe(_ tool: ToolContent) -> (label: String, detail: String?) {
        let title = (tool.state.title?.isEmpty == false) ? tool.state.title : nil
        let meta = tool.state.metadata

        func input(_ key: String) -> String? {
            guard let v = tool.state.input?[key]?.value else { return nil }
            let s = "\(v)"
            return s.isEmpty ? nil : s
        }

        let label = label(for: tool.tool)
        var detail: String?

        switch tool.tool.lowercased() {
        case "read":
            // Web shows just the filename (getFilename(input.filePath)).
            detail = base(title ?? input("filePath"))
        case "edit", "write", "patch", "apply_patch":
            detail = base(title ?? input("filePath"))
            if let badge = diffBadge(meta?.filediff) { detail = join(detail, badge) }
        case "bash", "shell":
            detail = title ?? input("description") ?? input("command")
            if let exit = meta?.exit, exit != 0 { detail = join(detail, "exit \(exit)") }
        case "grep":
            detail = title ?? input("pattern")
            if let n = meta?.matches {
                detail = join(detail, n == 0 ? "no matches" : "\(n) match\(n == 1 ? "" : "es")")
            }
        case "glob":
            detail = title ?? input("pattern")
        case "list":
            detail = base(title ?? input("path"))
        case "webfetch":
            detail = title ?? input("url")
        case "websearch":
            detail = title ?? input("query")
        case "todowrite", "todo":
            detail = todoRatio(meta?.todos) ?? title
        case "task", "question":
            detail = title
        default:
            detail = title
        }
        return (label, detail)
    }

    /// A tool's output stripped of OpenCode's LLM-facing XML wrapper, so the
    /// detail screen shows the real content (code, listing, …) instead of raw
    /// `<path>…</path><type>…</type><content>…</content>` tags — matching the web
    /// client, which renders the inner content, not the envelope.
    static func cleanOutput(_ tool: ToolContent) -> String? {
        guard let output = tool.state.output?.trimmingCharacters(in: .whitespacesAndNewlines),
              !output.isEmpty else { return nil }
        // read wraps a file as <content>…</content> and a directory as
        // <entries>…</entries> (both after <path>/<type>). Show the inner payload.
        for tag in ["content", "entries"] {
            if let inner = between(output, open: "<\(tag)>", close: "</\(tag)>") {
                return inner.trimmingCharacters(in: .newlines)
            }
        }
        return output
    }

    /// Text strictly between the first `open` and the following `close` marker.
    private static func between(_ s: String, open: String, close: String) -> String? {
        guard let o = s.range(of: open),
              let c = s.range(of: close, range: o.upperBound..<s.endIndex) else { return nil }
        return String(s[o.upperBound..<c.lowerBound])
    }

    /// "+N −M" badge from an edit/write file diff, or nil when both are zero.
    /// Uses a real minus sign (U+2212), matching the web client's DiffChanges.
    static func diffBadge(_ diff: FileDiff?) -> String? {
        guard let diff, (diff.additions ?? 0) != 0 || (diff.deletions ?? 0) != 0 else { return nil }
        return "+\(diff.additions ?? 0) −\(diff.deletions ?? 0)"
    }

    /// "completed/total" for a todowrite, mirroring the web client's "X/Y".
    static func todoRatio(_ todos: [MetaTodo]?) -> String? {
        guard let todos, !todos.isEmpty else { return nil }
        let done = todos.filter { $0.status == "completed" }.count
        return "\(done)/\(todos.count)"
    }

    /// Last path component, matching the web client's getFilename helper.
    static func base(_ path: String?) -> String? {
        guard let path, !path.isEmpty else { return nil }
        return (path as NSString).lastPathComponent
    }

    private static func join(_ lhs: String?, _ rhs: String) -> String {
        guard let lhs, !lhs.isEmpty else { return rhs }
        return lhs + "  " + rhs
    }
}

/// Compact presentation of a `patch` part — the files changed in one turn. The
/// standalone patch part carries only the file list (no counts), matching what
/// the web client renders without the per-file diff metadata.
enum PatchDisplay {
    static func summary(_ patch: PatchContent) -> String {
        let n = patch.files.count
        return "\(n) file\(n == 1 ? "" : "s")"
    }
}

/// Compact presentation of a `file` part (IDE context attachment): the filename
/// plus the single line / range parsed from the `url` query (`?start=&end=`).
enum FileRefDisplay {
    static func chip(_ part: FileRefContent) -> String {
        let name = part.filename ?? lastComponent(part.url) ?? "file"
        guard let range = lineRange(part.url) else { return name }
        return "\(name):\(range)"
    }

    /// "266" for start==end, "266-280" for a span, nil when no range is encoded.
    static func lineRange(_ url: String?) -> String? {
        guard let url, let comps = URLComponents(string: url) else { return nil }
        let items = comps.queryItems ?? []
        let start = items.first { $0.name == "start" }?.value
        let end = items.first { $0.name == "end" }?.value
        guard let start, !start.isEmpty else { return nil }
        if let end, !end.isEmpty, end != start { return "\(start)-\(end)" }
        return start
    }

    private static func lastComponent(_ url: String?) -> String? {
        guard let url, let comps = URLComponents(string: url), !comps.path.isEmpty else { return nil }
        return (comps.path as NSString).lastPathComponent
    }
}

File: packages/android/app/src/main/java/studio/eugenezakharov/opencode/ui/session/ToolDisplay.kt
package studio.eugenezakharov.opencode.ui.session

import studio.eugenezakharov.opencode.api.models.PartContent

/**
 * Compact, web-client-style presentation of a message's non-text parts: a verb
 * label plus a concise target (file / command / pattern) and an optional badge
 * (diff counts, match count). Pure logic so it can be unit-tested against real
 * session data — the exact Kotlin twin of the iOS `ToolDisplay`, mirroring
 * `getToolInfo()` in the OpenCode web client.
 */
object ToolDisplay {
    /** The verb shown for a tool, matching the web client's English labels. */
    fun label(tool: String): String = when (tool.lowercase()) {
        "read" -> "Read"
        "list" -> "List"
        "glob" -> "Glob"
        "grep" -> "Grep"
        "bash", "shell" -> "Shell"
        "edit" -> "Edit"
        "write" -> "Write"
        "patch", "apply_patch" -> "Patch"
        "webfetch" -> "Webfetch"
        "websearch" -> "Web Search"
        "task" -> "Agent"
        "todowrite", "todo" -> "To-dos"
        "question" -> "Questions"
        else -> tool
    }

    /**
     * A tool's verb plus a one-line target + badge, e.g.
     * `"Edit" to "ast_query.h  +5 −4"`, `"Grep" to "Foo|Bar  no matches"`.
     */
    fun describe(tool: PartContent.Tool): Pair<String, String?> {
        val title = tool.title?.takeIf { it.isNotEmpty() }
        val meta = tool.metadata
        fun input(key: String) = tool.input[key]?.takeIf { it.isNotEmpty() }

        val label = label(tool.tool)
        val detail: String? = when (tool.tool.lowercase()) {
            "read" -> base(title ?: input("filePath"))
            "edit", "write", "patch", "apply_patch" ->
                join(base(title ?: input("filePath")), diffBadge(meta))
            "bash", "shell" -> {
                val d = title ?: input("description") ?: input("command")
                val exit = meta?.exit
                if (exit != null && exit != 0) join(d, "exit $exit") else d
            }
            "grep" -> {
                val d = title ?: input("pattern")
                val n = meta?.matches
                if (n != null) join(d, if (n == 0) "no matches" else "$n match${if (n == 1) "" else "es"}") else d
            }
            "glob" -> title ?: input("pattern")
            "list" -> base(title ?: input("path"))
            "webfetch" -> title ?: input("url")
            "websearch" -> title ?: input("query")
            "todowrite", "todo" -> todoRatio(meta) ?: title
            "task", "question" -> title
            else -> title
        }
        return label to detail
    }

    /**
     * A tool's output stripped of OpenCode's LLM-facing XML wrapper, so the
     * detail screen shows the real content (code, listing, …) instead of raw
     * `<path>/<type>/<content>` tags — matching the web client and the iOS twin.
     */
    fun cleanOutput(tool: PartContent.Tool): String? {
        val output = tool.output?.trim()?.takeIf { it.isNotEmpty() } ?: return null
        return unwrapEnvelope(output) ?: output
    }

    /**
     * OpenCode's `read` tool wraps its result (for the LLM) as
     * `<path>…</path>\n<type>…</type>\n<WRAPPER>\n…payload…\n</WRAPPER>` — a file's
     * `<content>`, a directory's `<entries>`, etc. Return the inner payload of
     * whatever wrapper follows `<type>`, so any current/future wrapper is handled
     * without per-tag special cases (server: packages/opencode/src/tool/read.ts).
     */
    private fun unwrapEnvelope(s: String): String? {
        if (!s.startsWith("<path>")) return null
        val afterType = s.indexOf("</type>").takeIf { it >= 0 } ?: return null
        val rest = s.substring(afterType + "</type>".length).trim()
        if (!rest.startsWith("<")) return null
        val gt = rest.indexOf('>').takeIf { it > 1 } ?: return null
        val name = rest.substring(1, gt)
        if (name.isEmpty() || name.contains(' ') || name.contains('/')) return null
        val close = rest.lastIndexOf("</$name>").takeIf { it > gt } ?: return null
        return rest.substring(gt + 1, close).trim('\n')
    }

    /** "+N −M" badge from an edit/write file diff, or null when both are zero. U+2212 minus. */
    fun diffBadge(meta: studio.eugenezakharov.opencode.api.models.ToolMeta?): String? {
        val a = meta?.additions ?: 0
        val d = meta?.deletions ?: 0
        if (a == 0 && d == 0) return null
        return "+$a −$d"
    }

    /** "completed/total" for a todowrite, mirroring the web client's "X/Y". */
    fun todoRatio(meta: studio.eugenezakharov.opencode.api.models.ToolMeta?): String? {
        val total = meta?.todoTotal ?: return null
        if (total == 0) return null
        return "${meta.todoCompleted ?: 0}/$total"
    }

    /** Last path component, matching the web client's getFilename helper. */
    fun base(path: String?): String? =
        path?.takeIf { it.isNotEmpty() }?.substringAfterLast('/')?.substringAfterLast('\\')

    private fun join(lhs: String?, rhs: String?): String? {
        if (rhs.isNullOrEmpty()) return lhs
        if (lhs.isNullOrEmpty()) return rhs
        return "$lhs  $rhs"
    }
}

/** Compact presentation of a `patch` part — the files changed in one turn. */
object PatchDisplay {
    fun summary(patch: PartContent.Patch): String {
        val n = patch.files.size
        return "$n file${if (n == 1) "" else "s"}"
    }
}

/**
 * Compact presentation of a `file` part (IDE context attachment): the filename
 * plus the line / range parsed from the `url` query (`?start=&end=`).
 */
object FileRefDisplay {
    fun chip(file: PartContent.FileRef): String {
        val name = file.filename ?: lastComponent(file.url) ?: "file"
        val range = lineRange(file.url) ?: return name
        return "$name:$range"
    }

    /** "266" for start==end, "266-280" for a span, null when no range is encoded. */
    fun lineRange(url: String?): String? {
        val query = url?.substringAfter('?', "")?.takeIf { it.isNotEmpty() } ?: return null
        var start: String? = null
        var end: String? = null
        for (pair in query.split('&')) {
            val kv = pair.split('=', limit = 2)
            if (kv.size != 2) continue
            when (kv[0]) {
                "start" -> start = kv[1]
                "end" -> end = kv[1]
            }
        }
        val s = start?.takeIf { it.isNotEmpty() } ?: return null
        return if (!end.isNullOrEmpty() && end != s) "$s-$end" else s
    }

    private fun lastComponent(url: String?): String? =
        url?.substringBefore('?')?.substringAfterLast('/')?.takeIf { it.isNotEmpty() }
}
